File: typescript/Behavioral/13-ChainOfResponsibility/Handler/BaseApprovalHandler.ts
import { IApprovalHandler } from "./IApprovalHandler";
import { ApprovalRequest } from "../Model/ApprovalRequest";

export abstract class BaseApprovalHandler implements IApprovalHandler {
  private next?: IApprovalHandler;

  setNext(next: IApprovalHandler): IApprovalHandler {
    this.next = next;
    return next;
  }

  handle(request: ApprovalRequest): void {
    if (this.next) this.next.handle(request);
    else console.log(`  ❌ No handler could approve: ${request}`);
  }
}

File: csharp/Behavioral/13-ChainOfResponsibility/Handler/BaseApprovalHandler.cs
public abstract class BaseApprovalHandler : IApprovalHandler
{
    private IApprovalHandler? _next;

    public IApprovalHandler SetNext(IApprovalHandler next)
    {
        _next = next;
        return next;
    }

    public virtual void Handle(ApprovalRequest request)
    {
        if (_next is not null)
            _next.Handle(request);
        else
            Console.WriteLine($"  ❌ No handler could approve: {request}");
    }
}
